server/ARCHITECTURE.md

```diff
@@ -0,0 +1,73 @@
+# Architecture
+
+This document describes the high-level architecture of Waterpark. If you want to familiarize yourself
+with the codebase, you are just in the right place!
+
+### Bird's Eye View
+
+![](https://blog.cleancoder.com/uncle-bob/images/2012-08-13-the-clean-architecture/CleanArchitecture.jpg)
+
+Waterpark's architecture is heavily modelled with Domain Driven Design. DDD is a strong tool when modelling complex and rich domains. You can learn more about Loo Labs' motivations for DDD [here]().
+
+On the highest level, you can think of Domain Driven Design as encapsulating pure functional
+business logic in the deepest layer of a system. This way, the business logic, or _domain layer_
+does not know anything about the outside world. It has no idea about how to talk to clients via
+express. Nor does it have a clue on how to talk to databases via ORM. This is great because
+the domain layer is decoupled from the _routing_ and _persistence_ layers. Decoupling layers gives
+us two main advantages!
+
+1. Maintainability
+2. Testing
+
+**Maintainabilitiy**
+
+If we want to rip out express for another http library, we can do so with ease. All we have to do
+is change the code in our routing layer, without touching _any_ of our important business logic, which is protected in our _domain_ layer. The same thing can be said on the topic of ripping one ORM for the other.
+
+**Testing**
+
+The main thing you want to test is your business logic. The main way to test functionality is with
+unit tests. Integration tests, on the other hand, ensure that all units work in harmony together. Wouldn't it suck if running unit tests required spinning up expensive infrastructure like express and postgres servers? All we want to do is test the functionality of business logic! That's a lot of time wasted both by the developer when running tests locally and by the CI/CD servers when running tests in our release pipeline.
+
+But, the beautiful thing with DDD is that our business logic is encapsulated in a pure functional
+layer which only consists of data structures and algorithms. This allows us to run unit tests and
+verify the correctness of our system (unit-wise) without spinning up _any_ infrastructure.
+
+If the concepts are still hazy at this point don't worry! Things will clear up after reading the next
+section.
+
+### Code Map
+
+![](https://i.imgur.com/LKRVNIH.png)
+The code map will list important files, modules, and types. There will be no direct links in this document however — this is because links often go stale. Instead, use symbol search to find the mentioned entities by name!
+
+#### `modules`
+
+This is the main directory which contains all the server code. The following directories we'll talk
+about can be found under any module in _this_ directory.
+
+#### `modules/users/infra`
```

Comment:
Are we focusing on `users` as a concrete example?

Reply:
I think it's useful because of the picture of the codebase includes users to be expanded. Should we make it more explicit that `users` is an example?

Reply:
it'd be nice but I don't have any good format in mind for it; we're busy so let's defer this until the next time we touch this doc

Reply:
agreed